File: LeetCode/238. Product of Array Except Self.py
# Author : @AdarshGaur

class Solution:
    def productExceptSelf(self, nums: List[int]) -> List[int]:
        ans = [1] * len(nums)
        pref = suf = 1
        for i in range(len(nums)):
            ans[i] *= pref
            ans[-1-i] *= suf
            pref *= nums[i]
            suf *= nums[-1-i]
        
        return ans;

File: LeetCode/238. Product of Array Except Self.java
class Solution {
    public int[] productExceptSelf(int[] nums) {
        int n = nums.length;
        int prefix = 1, suffix = 1;
        int[] ans = new int[n];
        Arrays.fill(ans, 1);
        for(int i=0; i<n; i++){
            // keep the prefix
            ans[i] *= prefix;
            prefix *= nums[i];

            // keep the suffix
            ans[n-1 -i] *= suffix;
            suffix *= nums[n-1 -i];
        }
        return ans;
    }
}
